src/extension/serializer.ts

```diff
@@ -844,43 +854,302 @@ export class GrpcSerializer extends SerializerBase {
+    if (sid && docUri) {
+      const relativePath = path.basename(docUri.fsPath)
+      session = {
+        id: sid,
+        document: { relativePath },
+      } as es_proto.RunmeSession
+    }
+
+    const outputs = { enabled: true, summary: true }
+    const options = {
+      outputs,
+      session,
+    } as es_proto.SerializeRequestOptions
+
+    maskedNotebook.cells.forEach((cell) => {
+      cell.value = maskString(cell.value)
+      cell.outputs.forEach((out) => {
+        out.items.forEach((item) => {
+          if (item.mime === OutputType.stdout) {
+            const outDecoded = Buffer.from(item.data).toString('utf8')
+            item.data = Buffer.from(maskString(outDecoded))
+          }
+        })
+      })
+    })
+
+    const plainReq = this.getSerializationRequest({ notebook, options })
+    const plainRes = this.client!.serialize(plainReq)
+
+    const maskedReq = this.getSerializationRequest({ notebook: maskedNotebook, options })
+    const masked = this.client!.serialize(maskedReq).then((response) => {
+      if (response.result === undefined) {
+        console.error('serialization of masked notebook failed')
+        return Promise.resolve(new Uint8Array())
+      }
+      return response.result
+    })
+
+    if (!cacheId) {
+      console.error('skip masked caching since no lifecycleId was found')
+    } else {
+      this.maskedCache.set(cacheId, masked)
+    }
+
+    const plain = await plainRes
+    if (plain.result === undefined) {
+      throw new Error('serialization of notebook outputs failed')
+    }
+
+    const bytes = plain.result
+    if (!cacheId) {
+      console.error('skip plain caching since no lifecycleId was found')
+    } else {
+      this.plainCache.set(cacheId, Promise.resolve(bytes))
+    }
+
+    await Promise.all([plain, masked])
+  }
+
+  protected async reviveNotebook(
+    content: Uint8Array,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    token: CancellationToken,
+  ): Promise<Serializer.Notebook> {
+    // const markdown = Buffer.from(content).toString('utf8')
+    const deserializeRequest = new es_proto.DeserializeRequest()
+    deserializeRequest.source = content
+    let res
+    try {
+      res = await this.client.deserialize(deserializeRequest)
+    } catch (e: any) {
+      if (e.name === 'ConnectError') {
+        e.message = `Unable to connect to the serializer service at ${this.serializerServiceUrl}`
+      }
+      log.error('Error in reviveNotebook  ', e as any)
+      throw e
+    }
+    const notebook = res.notebook
+
+    if (!notebook) {
+      return this.printCell('⚠️ __Error__: no cells found!')
+    }
+    return notebook as any as Serializer.Notebook // ugly cast :(
+  }
+}
+
+class Marshal {
+  // convert vscode.NotebookData to one of two runme Notebook proto type bindings
+  public static notebook(
+    data: NotebookData,
+    notebookProto: es_proto.Notebook | Notebook,
```

Comment:
Is there a way to reuse code without taking a union type? What is the code doing that it can take either type?

Reply:
I think extracting the code that rely on particulars of each proto transpiler type, within it's host class, and re-incorporating the methods currently contained in `Marshal` will eliminate this union type.  The proto types will then be specific to that class (for example: `protoNotebookType` as it exist differently in `GrpcSerializer` and `ConnectSerializer`, and would thus be used as the type in each classes method signatures).  The union type were required here because there were two serializers that called it, each w/ their own proto types.